presto-main/src/main/java/com/facebook/presto/operator/PagesSpatialIndexSupplier.java

```diff
@@ -137,7 +144,15 @@ private static STRtree buildRTree(LongArrayList addresses, List<List<Block>> cha
                 partition = toIntExact(INTEGER.getLong(partitionBlock, blockPosition));
             }
 
-            rtree.insert(getJtsEnvelope(ogcGeometry, radius), new GeometryWithPosition(ogcGeometry, partition, position));
+            GeometryWithPosition geometryWithPosition = new GeometryWithPosition(ogcGeometry, partition, position);
+            rtree.insert(getJtsEnvelope(ogcGeometry, radius), geometryWithPosition);
+
+            estimatedLeafNodesSize = estimatedLeafNodesSize + ENVELOPE_INSTANCE_SIZE + geometryWithPosition.getEstimatedMemorySizeInBytes();
+
+            // update memory size when 1000 more leaf nodes are inserted
+            if (position % 1000 == 999) {
+                localUserMemoryContext.setBytes(estimatedLeafNodesSize);
+            }
         }
 
         rtree.build();
```

Comment:
What about the trailing geometries after the last localUserMemoryContext, and the cost of the built rtree?  For the JTS rtree this is quite significant.

Reply:
The overhead of non-leaf node is about 10% of leaf nodes. After SRTree is built, a final size check is already done on all nodes at all levels, which includes trailing geometries.